### ISSUE TITLE
Make maintainableartefacts that are external_references resolvable as in v0.9

### ISSUE THREAD

**Issue by dr-leo** [dr-leo/pandaSDMX#156]
In v0.9, a maintainableartefact or even nameableartefact whose is_external_reference attrib is True, is represented in the XML msg as Ref node. It has some XML attributes allowing to prepare a request for the referenced artefact. In v0.9, such artefacts were represented by a Ref object which did not fully correspond to the IM. Yet, the Ref.resolve() method provided a convenient way to request the referenced object. v1.0dev requires the user to construct a request from scratch. This is not the intention of the XML representation (in particular its package attrib).

**Comment by khaeru** [comment]
Just to state this differently to be sure I understand…currently, the user can do this:
```python
import pandasdmx as sdmx

ESTAT = sdmx.Request('ESTAT')
data_msg = estat.data('une_rt_a', params={'startPeriod': '2017'})
dsd_0 = data_msg.data[0].structured_by  # is_external_reference=True; unpopulated
dsd_1 = ESTAT.get(resource=dsd_0)  # fully populated
```
Suppose the user throws away the Request object. They can do:
```
del ESTAT
dsd_1 = sdmx.Request('ESTAT').get(resource=dsd_0)
```
The proposed enhancement, if I understand, is that the user can do this without having to recall that dsd_0 was retrieved from ESTAT, e.g.:
```
dsd_1 = dsd_0.resolve()
```

If so, then I can think of at least two ways to implement:
1. Use the URN (so this would follow #154), for instance: https://github.com/dr-leo/pandaSDMX/blob/c578985ac97ab358c3866de0763e2db6b580c327/pandasdmx/tests/data/ECB_EXR/1/M.USD.EUR.SP00.A.xml#L7-L11
   ...and the regex I already added to parse these: https://github.com/dr-leo/pandaSDMX/blob/c578985ac97ab358c3866de0763e2db6b580c327/pandasdmx/reader/sdmxml.py#L39-L44
   …and then use the 'agency' group to instantiate the Request object.
2. For MaintainableArtefacts, only when is_external_reference=True, store a private attribute with the (pandaSDMX internal) name of the data source used, i.e. corresponding to sources.json.

Either way, .`resolve()` could be pretty short.

However, there are ambiguous situations, and I think a clear design is needed before writing any code:

The URN mentions the agency that is the **data provider**. But one might get a MaintainableArtefact reference from to an object where a second, different agency that is the **data source**.
- Example A: fetch a message from ESTAT that has an external reference to a code list from the SDMX Global Registry (SGR). In this case, implementation (1) above might work, but (2) will not work if ESTAT data source does not republish the metadata from SGR data provider.
- Example B: fetch a message from data source FOO that republishes data from ESTAT, but the data reference structures that are published by data provider BAR. The external references could be resolved by querying BAR, FOO, or ESTAT.

**Comment by khaeru** [comment]
To add:
- If https://github.com/dr-leo/pandaSDMX/issues/156#issuecomment-609014542 is correct, the issue title should be "Add MaintainableArtefact.resolve()". External references to MaintainableArtefacts *are* already resolvable—so long as one doesn't throw away the Request object used to get the objects in the first place.
- Since they are resolvable, what was possible in 0.9 is still possible in 1.0. The code `ESTAT.get(resource=dsd_0) or `sdmx.Request('ESTAT').get(resource=dsd_0)` may be slightly longer than `dsd_0.resolve()`, but it works.
- I would suggest we update the docs (e.g. at https://pandasdmx.readthedocs.io/en/master/howto.html#use-the-references-query-parameter) to explain how to do this using my snippets from above, and retarget this issue for 1.1, to give time to discuss the intended behaviour.

**Comment by dr-leo** [comment]
Agreed that a resolve method could wait until v1.1. However, if the concerns re. ambiguous situations (examples a) and b)) are  justified, then the proposed way of resolving by passing the external ref DSD to Request.get is also not guaranteed to work.

I think the issue comes down to understanding whether the IM or some tech note forces data sources and provider to avoid such ambiguities. If so, a resolve method would be pretty short indeed. It could even be done with a one-off request object with no agency (call this anonyous request obj) and pass it an URL or use sdmx.read_url().

**Comment by dr-leo** [comment]
To ad:

1. Isn't it easier to use URI rather than URN to retrieve the referenced resource?
1. Note that in v0.9, `Ref.resolve()` first checks whether the target sits in the same message in which case the target is returned. E.g., a Categorization in essence references the categorized artefact and one or more categories. Query the list of dataflows from ECB with Default value for 'references' (this time the Request.get() arg) to see this in Action. Conversely, if the target is not in the same message, it is optionally fetched remotely. E.g., structured_by attrib of a dataset. 

In summary, the added value of a `resolve()`method (wherever it should be attached to in the Absence of a model.Ref class) would be to provide a Standard way to check if the target is already present in Memory or must be queried remotely. The best way to deploy such method could be a dedicated `Ref` mixin class or Decorator. 

I haven't checked how v1.0dev implements those Things. But a single method for doing it in all cases where Refs occur in the XML or SDMXJSON in the future would help avoid Code duplication and ad hoc Solutions.

**Comment by khaeru** [comment]
> Isn't it easier to use URI rather than URN to retrieve the referenced resource?

Definitely, but as we saw in #154, #161, they aren't often supplied. I guess if they were, that would be the first location to try.

> 2. Note that in v0.9, `Ref.resolve()` first checks whether the target sits in the same message in which case the target is returned. E.g., a Categorization in essence references the categorized artefact and one or more categories.

Basically, the 1.0 SDMX-ML is to use a couple of indexes and methods like `Reader._maintained()` to **always** get an object that's in the same message, providing it exists. So we can basically say that `resolve()` would *only* ever need to be called when the MaintainableArtefact in question is not in the current message (or a DSD supplied to help parse it).

The phrase "already in memory" suggests something a little more ambitious. Suppose you do this:
```
msg1 = sdmx.read_sdmx('msg1.xml')
msg2 = sdmx.read_sdmx('msg2.xml')
```
When parsing msg2.xml, the Reader has no way to be aware that msg1 (or its contents) exists, and vice versa. For it to know that—without passing references—would require some kind of global (i.e. process-wide) index of objects, maybe by URN/URI. That's not impossible, but it's definitely a new feature.

> Agreed that a resolve method could wait until v1.1.

…so I'll mark this 'enhancement' and set to the 1.1 milestone.

**Comment by dr-leo** [comment]
Not really needed as reference can be passed to Request.get
